// @fortaine/openai/dist/http/isomorphic-fetch.js
import { ResponseContext } from './http.js';
import { from } from '../rxjsStub.js';
import createFetch from '@fortaine/fetch';
export class IsomorphicFetchHttpLibrary {
    static fetch(...args) {
        const fetcher = createFetch();
        return fetcher(...args);
    }
    send(request) {
        let method = request.getHttpMethod().toString();
        let body = request.getBody();
        const resultPromise = IsomorphicFetchHttpLibrary.fetch(request.getUrl(), {
            method: method,
            body: body,
            headers: request.getHeaders(),
        }).then((resp) => {
            const headers = {};
            resp.headers.forEach((value, name) => {
                headers[name] = value;
            });
            const body = {
                text: () => resp.text(),
                binary: () => resp.buffer(),
                stream: () => resp.body
            };
            return new ResponseContext(resp.status, headers, body);
        });
        return from(resultPromise);
    }
}
